RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -566,69 +569,51 @@ void L2TauNNProducer::fillCaloRecHits(tensorflow::Tensor& cellGridMatrix,
-  VtxGood.reserve(nv);
-
-  std::vector<double> maxChi2_;
-  std::vector<double> pTSquaredSum(nv);
-
-  for (int j = nv - 1; j >= 0; --j) {
-    std::vector<int> trk_ass_to_vtx;
-    auto vtx_idx = patavtx_soa.sortInd[j];
-    assert(vtx_idx < nv);
-    for (int trk_idx = 0; trk_idx < maxTracks; trk_idx++) {
-      int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track == int16_t(vtx_idx))
-        trk_ass_to_vtx.push_back(trk_idx);
-    }
-    auto nt = trk_ass_to_vtx.size();
-    if (nt == 0) {
-      continue;
-    }
-    if (nt < 2) {
-      trk_ass_to_vtx.clear();
-      continue;
+  trkGood.clear();
+  trkGood.reserve(maxTracks);
+  vtxGood.clear();
+  vtxGood.reserve(nv);
+  auto const* quality = patatracks_tsoa.qualityData();
+
+  // No need to sort either as the algorithms is just using the max (not even the location, just the max value of pt2sum).
+  std::vector<float> pTSquaredSum(nv, 0);
+  std::vector<int> nTrkAssociated(nv, 0);
+
+  for (int32_t trk_idx = 0; trk_idx < maxTracks; ++trk_idx) {
+    auto nHits = patatracks_tsoa.nHits(trk_idx);
+    if (nHits == 0) {
+      break;
     }
-    for (const auto& trk_idx : trk_ass_to_vtx) {
-      int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track != vtx_idx)
-        continue;
-      double patatrackPt = patatracks_tsoa.pt[trk_idx];
-      if (patatrackPt < trackPtMin_)
-        continue;
-      if (patatracks_tsoa.chi2(trk_idx) > trackChi2Max_)
-        continue;
-      if (patatrackPt > trackPtMax_) {
-        patatrackPt = trackPtMax_;
+    int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
+    if (vtx_ass_to_track >= 0 && vtx_ass_to_track < nv) {
+      auto patatrackPt = patatracks_tsoa.pt[trk_idx];
+      ++nTrkAssociated[vtx_ass_to_track];
+      if (patatrackPt >= trackPtMin_ && patatracks_tsoa.chi2(trk_idx) <= trackChi2Max_) {
+        patatrackPt = std::min(patatrackPt, trackPtMax_);
+        pTSquaredSum[vtx_ass_to_track] += patatrackPt * patatrackPt;
       }
-      pTSquaredSum.at(vtx_idx) += patatrackPt * patatrackPt;
     }
-  }
-  std::vector<size_t> sortIdxs(nv);
-  std::iota(sortIdxs.begin(), sortIdxs.end(), 0);
-  std::sort(sortIdxs.begin(), sortIdxs.end(), [&](size_t const i1, size_t const i2) {
-    return pTSquaredSum[i1] > pTSquaredSum[i2];
-  });
-  auto const minFOM_fromFrac = pTSquaredSum[sortIdxs.front()] * fractionSumPt2_;
-
-  for (int j = nv - 1; j >= 0; --j) {
-    auto idx = patavtx_soa.sortInd[j];
-
-    if (VtxGood.size() >= maxVtx_) {
-      break;
+    if (nHits > 0 and quality[trk_idx] >= pixelTrack::Quality::loose) {
+      trkGood.push_back(trk_idx);
     }
-    if (pTSquaredSum[idx] >= minFOM_fromFrac && pTSquaredSum[idx] > minSumPt2_) {
-      VtxGood.push_back(idx);
+  }
+  if (nv > 0) {
+    const auto minFOM_fromFrac = (*std::max_element(pTSquaredSum.begin(), pTSquaredSum.end())) * fractionSumPt2_;
+    for (int j = nv - 1; j >= 0; --j) {
+      auto vtx_idx = patavtx_soa.sortInd[j];
```

Comment:
I'm generally confused about the role of `vtxGood`, but obviously I'm not the expert here.

Is the order of the elements in `vtxGood` important?
 - If not, why do you need to use the sorting from the input vertices (`soa.sortInd[j]`)?
 - If yes, since the sorting from the input vertices is used, shouldn't one use the f.o.m. that was used for that sorting? Is that the purpose of `pTSquaredSum` (to match exactly the f.o.m. used for the original sorting)? If so, wouldn't that value already be available in `soa.ptv2[sortInd[j]]`?

Is `vtxGood` meant to mimic the "trimmed" pixel vertices? If yes, why not simply use those as (additional) input to this producer, without having to recompute all this and having to keep this whole source code aligned with the pixel reco?

Reply:
> Is the order of the elements in `vtxGood` important?
> * If not, why do you need to use the sorting from the input vertices (`soa.sortInd[j]`)?

No, the order is not important. `patavtx_soa.sotrInd[vtx_idx]` is needed to make track-vertex association, because `patavtx_soa.idv[trk_idx]` refers to `sortInd` instead of `vtx_idx`.

> Is `vtxGood` meant to mimic the "trimmed" pixel vertices? If yes, why not simply use those as (additional) input to this producer, without having to recompute all this and having to keep this whole source code aligned with the pixel reco?

In `L2TauTag` producer we prefer to avoid the use of legacy formats to eliminate unnecessary overhead from conversion and creation of the new collection. While indeed in the current version, the vertex selection mimics "trimmed" vertex producer logic, we plan to change the selection for future training (e.g. use `soa.ptv2`, review cuts applied to `trkGood` and `vtxGood`).

Reply:
> No, the order is not important. [..]

In this particular loop, I don't see `.idv`, so it's not clear to me how [1] would differ from [2] if the order of `vtxGood` is not important.

[1]
```cpp
    for (int j = nv - 1; j >= 0; --j) {
      auto vtx_idx = patavtx_soa.sortInd[j];
      assert(vtx_idx < nv);
      if (nTrkAssociated[vtx_idx] >= 2 && pTSquaredSum[vtx_idx] >= minFOM_fromFrac &&
          pTSquaredSum[vtx_idx] > minSumPt2_) {
        vtxGood.push_back(vtx_idx);
      }
    }
```
[2]
```cpp
    for (int vtx_idx = nv - 1; vtx_idx >= 0; --vtx_idx) {
      if (nTrkAssociated[vtx_idx] >= 2 && pTSquaredSum[vtx_idx] >= minFOM_fromFrac &&
          pTSquaredSum[vtx_idx] > minSumPt2_) {
        vtxGood.push_back(vtx_idx);
      }
    }
```

Reply:
Sorry, I've misunderstood your original comment. Yes, in this particular loop [1] and [2] will result in equivalent NN inputs. So let's use [2] since it is simpler.

Reply:
Hm.. actually, as @valeriadamante pointed me out, after adding back `maxVtx_` [1] and [2] will no longer be equivalent because if the number of potentially good vertices is greater than `maxVtx_`, we would want to keep the most energetic ones. However, the order in which they will be stored in `vtxGood` is not important.

Reply:
Right, I see. In principle, one could also fill `vtxGood` then use `if (vtxGood.size > maxVtx_) vtxGood.resize(maxVtx_)` afterwards (without having the `if (vtxGood.size > maxVtx_)` check inside the `for` loop). I don't know what's better and faster, it might depend on the typical value of `vtxGood.size` and `maxVtx_`. You are free to use the implementation you prefer, of course.